### PR TITLE
Add support recoverable signatures.

### PR DESCRIPTION
**Pull request by gatoWololo**
Implements method to compute recoverable signatures. Closes Issue #508.

Tested by adding recoverable signature function call to sign unit tests.
Adds documentation for the threaded example CLI.

### CODE REVIEW

File: src/sign/interactive_sign/participant.rs [hunk 1]
```diff
@@ -488,6 +491,21 @@ mod tests {
             )
             .is_ok());
 
+        // Check we are able to create a recoverable signature.
```
Comment (gatoWololo) [inline src/sign/interactive_sign/participant.rs:494]: This code ensures we can successfully generate the recovery ID and derive the public key from that recovery ID on this unit test.

File: src/sign/non_interactive_sign/participant.rs [hunk 2]
```diff
@@ -626,6 +629,22 @@ mod test {
                 distributed_sig.as_ref()
             )
             .is_ok());
+
```
Comment (gatoWololo) [inline src/sign/non_interactive_sign/participant.rs:632]: This code ensures we can successfully generate the recovery ID and derive the public key from that recovery ID on this unit test.

**Comment by gatoWololo** [comment]
Some additional information for the reviewer. Please see #508 for the original idea. We purposely chose to implement this as an additional method on our `Signature` type. The alternative approach would have been change the output type on:
```rust
impl ProtocolParticipant for SignParticipant {
    type Input = Input;
    type Output = Signature;
```
From `Signature` to some new type, `RecoverableSignature`, this would have forced the caller to always create a `RecoverableSignature`. We like the generality of this library though, so we chose to still create a `Signature` and leave the recoverable ID as "opt-in"

**Review by becgabri — APPROVED** [review]
Yes, this all makes sense to me and it looks good. Just need to remember that when this gets called the recovery ID at least needs to be concatenated with the signature.

File: src/sign/interactive_sign/participant.rs (hunk 1 above)
Comment (gatoWololo) [reply src/sign/interactive_sign/participant.rs:494]: Nope, just wanted to point it out!

**Comment by gatoWololo** [comment]
> Just need to remember that when this gets called the recovery ID at least needs to be concatenated with the signature.

Right, I haven't seen a standardized format to include the signature + recovery ID. So I leave them as separate values so that the caller can choose how to handle this.

**Pull request merged**